{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AdaptiveFancyArrow', 'Copyable', 'File', 'InvalidFileError', 'TextFile', '__author__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', 'colors', 'copyable', 'file', 'sequence', 'set_font_size_in_coord', 'temp', 'temp_dir', 'temp_file', 'visualize']\n",
      "['Alphabet', 'AlphabetError', 'AlphabetMapper', 'AnnotatedSequence', 'Annotation', 'CodonTable', 'Feature', 'GeneralSequence', 'LetterAlphabet', 'Location', 'NucleotideSequence', 'ProteinSequence', 'Sequence', '__author__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'alphabet', 'annotation', 'codec', 'codon', 'find_subsequence', 'find_symbol', 'find_symbol_first', 'find_symbol_last', 'search', 'seqtypes', 'sequence']\n",
      "['A', 'C', 'G', 'T']\n",
      "['A', 'C', 'G', 'T']\n",
      "['__abstractmethods__', '__add__', '__class__', '__copy_create__', '__copy_fill__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_compl_dict', '_complement_func', '_dtype', '_key', '_key_code', '_val_code', '_value', 'alphabet', 'alphabet_amb', 'ambiguous_alphabet', 'code', 'compl_symbol_dict', 'complement', 'copy', 'get_alphabet', 'get_symbol_frequency', 'is_valid', 'reverse', 'symbols', 'translate', 'unambiguous_alphabet']\n",
      "{'A': 2, 'C': 2, 'G': 2, 'T': 3}\n",
      "TAGACGTAC\n",
      "CATGCAGAT\n",
      "CATGCAGAT\n",
      "MIT\n",
      "{'A': 0, 'C': 0, 'D': 0, 'E': 0, 'F': 0, 'G': 0, 'H': 0, 'I': 1, 'K': 0, 'L': 0, 'M': 1, 'N': 0, 'P': 0, 'Q': 0, 'R': 0, 'S': 0, 'T': 1, 'V': 0, 'W': 0, 'Y': 0, 'B': 0, 'Z': 0, 'X': 0, '*': 0}\n",
      "M MET\n",
      "I ILE\n",
      "T THR\n",
      "9\n",
      "2\n",
      "3\n",
      "[5]\n",
      "[2 5]\n",
      "2\n",
      "5\n",
      "ICM\n",
      "AAA K      AAC N      AAG K      AAT N\n",
      "ACA T      ACC T      ACG T      ACT T\n",
      "AGA R      AGC S      AGG R      AGT S\n",
      "ATA I      ATC I      ATG M i    ATT I\n",
      "\n",
      "CAA Q      CAC H      CAG Q      CAT H\n",
      "CCA P      CCC P      CCG P      CCT P\n",
      "CGA R      CGC R      CGG R      CGT R\n",
      "CTA L      CTC L      CTG L i    CTT L\n",
      "\n",
      "GAA E      GAC D      GAG E      GAT D\n",
      "GCA A      GCC A      GCG A      GCT A\n",
      "GGA G      GGC G      GGG G      GGT G\n",
      "GTA V      GTC V      GTG V      GTT V\n",
      "\n",
      "TAA *      TAC Y      TAG *      TAT Y\n",
      "TCA S      TCC S      TCG S      TCT S\n",
      "TGA *      TGC C      TGG W      TGT C\n",
      "TTA L      TTC F      TTG L i    TTT F\n",
      "AAA K      AAC N      AAG K      AAT N\n",
      "ACA T      ACC T      ACG T      ACT T\n",
      "AGA R      AGC S      AGG R      AGT S\n",
      "ATA I      ATC I      ATG M i    ATT I\n",
      "\n",
      "CAA Q      CAC H      CAG Q      CAT H\n",
      "CCA P      CCC P      CCG P      CCT P\n",
      "CGA R      CGC R      CGG R      CGT R\n",
      "CTA L      CTC L      CTG L      CTT L\n",
      "\n",
      "GAA E      GAC D      GAG E      GAT D\n",
      "GCA A      GCC A      GCG A      GCT A\n",
      "GGA G      GGC G      GGG G      GGT G\n",
      "GTA V      GTC V      GTG V      GTT V\n",
      "\n",
      "TAA *      TAC Y      TAG *      TAT Y\n",
      "TCA S      TCC S      TCG S      TCT S\n",
      "TGA *      TGC C      TGG W      TGT C\n",
      "TTA L      TTC F      TTG L      TTT F\n",
      "E\n",
      "('GAA', 'GAG')\n",
      "AAA K      AAC N      AAG K      AAT N\n",
      "ACA T      ACC T      ACG T      ACT T\n",
      "AGA R      AGC S      AGG R      AGT S\n",
      "ATA I i    ATC I i    ATG M i    ATT I\n",
      "\n",
      "CAA Q      CAC H      CAG Q      CAT H\n",
      "CCA P      CCC P      CCG P      CCT P\n",
      "CGA R      CGC R      CGG R      CGT R\n",
      "CTA L      CTC L      CTG L i    CTT L\n",
      "\n",
      "GAA E      GAC D      GAG E      GAT D\n",
      "GCA A      GCC A      GCG A      GCT A\n",
      "GGA G      GGC G      GGG G      GGT G\n",
      "GTA V      GTC V      GTG V i    GTT V\n",
      "\n",
      "TAA *      TAC Y      TAG *      TAT Y\n",
      "TCA S      TCC S      TCG S      TCT S\n",
      "TGA *      TGC C      TGG W      TGT C\n",
      "TTA L      TTC F      TTG L i    TTT F\n",
      "AAA K      AAC N      AAG K      AAT N\n",
      "ACA T      ACC T      ACG T      ACT T\n",
      "AGA R      AGC S      AGG R      AGT S\n",
      "ATA M i    ATC I      ATG M i    ATT I\n",
      "\n",
      "CAA Q      CAC H      CAG Q      CAT H\n",
      "CCA P      CCC P      CCG P      CCT P\n",
      "CGA R      CGC R      CGG R      CGT R\n",
      "CTA T      CTC T      CTG T      CTT T\n",
      "\n",
      "GAA E      GAC D      GAG E      GAT D\n",
      "GCA A      GCC A      GCG A      GCT A\n",
      "GGA G      GGC G      GGG G      GGT G\n",
      "GTA V      GTC V      GTG V      GTT V\n",
      "\n",
      "TAA *      TAC Y      TAG *      TAT Y\n",
      "TCA S      TCC S      TCG S      TCT S\n",
      "TGA W      TGC C      TGG W      TGT C\n",
      "TTA L      TTC F      TTG L      TTT F\n"
     ]
    }
   ],
   "source": [
    "#With Biotite can create own custom alphabet\n",
    "#Load Biotite and matplotlib packages\n",
    "import biotite\n",
    "import biotite.sequence as seq\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#Similar to Biopython\n",
    "\n",
    "#Methods and features\n",
    "print(dir(biotite))\n",
    "#Methods and features in biotite.sequence\n",
    "print(dir(biotite.sequence))\n",
    "\n",
    "#Create dna sequence\n",
    "dna = seq.NucleotideSequence(\"ATCTGCATG\")\n",
    "\n",
    "#Get letters in sequence\n",
    "print(dna.alphabet)\n",
    "print(dna.get_alphabet())\n",
    "\n",
    "#Methods inside seq.NucleotideSequence\n",
    "print(dir(seq.NucleotideSequence))\n",
    "\n",
    "#Frequency of nucleotides within sequence\n",
    "print(dna.get_symbol_frequency())\n",
    "\n",
    "#Complement and reverse complement DNA structure \n",
    "#Native reverse complement method not in Biotite unlike in Biopython\n",
    "print(dna.complement())\n",
    "print(dna.reverse().complement())\n",
    "\n",
    "#Custom reverse complment function\n",
    "def reverse_complement(dna):\n",
    "    return seq.NucleotideSequence(str(dna)[::-1]).complement()\n",
    "print(reverse_complement(dna))\n",
    "\n",
    "#Create protein sequence\n",
    "protein_seq = seq.ProteinSequence(\"MIT\")\n",
    "print(protein_seq)\n",
    "\n",
    "#Get count/frequency of each of the 20 amino acids\n",
    "print(protein_seq.get_symbol_frequency())\n",
    "\n",
    "#Convert amino acid sequence from one to three letter codons\n",
    "for aa in protein_seq:\n",
    "    print(aa,seq.ProteinSequence().convert_letter_1to3(aa))\n",
    "\n",
    "#Convert 3 letter codon to 1 letter codon, most likely a bug is present here due to KeyError: 'M'\n",
    "#Not a big problem becasue for this functionality can use Biopython if needed\n",
    "'''protein_seq2 = seq.ProteinSequence(\"MetIleThr\")\n",
    "print(protein_seq2)\n",
    "for aa in protein_seq2:\n",
    "    print(aa,seq.ProteinSequence().convert_letter_3to1(aa))'''\n",
    "\n",
    "#Sequence manipulation, always convert sequence to string\n",
    "#Length of DNA\n",
    "print(len(dna))\n",
    "#Find the position of a nucleotide when it first appears\n",
    "print(str(dna).find(\"C\"))\n",
    "#Find the frequency of a nucleotide in sequence\n",
    "print(str(dna).count(\"T\"))\n",
    "\n",
    "#Search for a subsequence\n",
    "sub_seq = seq.NucleotideSequence(\"CAT\")\n",
    "print(seq.find_subsequence(dna,sub_seq))\n",
    "\n",
    "#Find all occurances of nucleotide in sequence and returns positions in sequence\n",
    "print(seq.find_symbol(dna,\"C\"))\n",
    "\n",
    "#Finds first occurance of nucleotide in sequence and returns position in sequence\n",
    "print(seq.find_symbol_first(dna,\"C\"))\n",
    "\n",
    "#Find last occurance of nucleotide in sequence and returns position in sequence\n",
    "print(seq.find_symbol_last(dna,\"C\"))\n",
    "\n",
    "#Plot of frequency of nucleotides\n",
    "freq = dna.get_symbol_frequency()\n",
    "#plt.bar(freq.keys(),freq.values())\n",
    "\n",
    "\n",
    "#Protein Synthesis\n",
    "\n",
    "#DNA to mRNA, mRNA to aa sequence\n",
    "#No native transcribe function unlike in biopython but can directly convert to aa sequence\n",
    "#May only return empty lists if DNA sequence not a multiple of 3\n",
    "#Also must specify complete=True then sequence will be returned as stop codon is issued\n",
    "protein = dna.translate(complete=True)\n",
    "print(protein)\n",
    "\n",
    "#Can check for all amino acids in CodonTable and other tables available for different species\n",
    "#Parsing argument as 1 shows official NCBI CodonTable\n",
    "print(seq.CodonTable.load(1))\n",
    "#NCBI Biotite CodonTable\n",
    "print(seq.CodonTable.default_table())\n",
    "\n",
    "#Search for a particular amino acid-nucleotide combinations\n",
    "table = seq.CodonTable.default_table()\n",
    "#Example find aa by codon\n",
    "print(table[\"GAA\"])\n",
    "#Example find codon from aa\n",
    "print(table[\"E\"])\n",
    "\n",
    "#CodonTable for Bacterial Genome using number\n",
    "bact_table = seq.CodonTable.load(11)\n",
    "print(bact_table)\n",
    "\n",
    "#CodonTable for Yeast Genome using name\n",
    "yeast_table = seq.CodonTable.load(\"Yeast Mitochondrial\")\n",
    "print(yeast_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
