{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Alignment', 'SubstitutionMatrix', '__author__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'align_multiple', 'align_optimal', 'align_ungapped', 'alignment', 'get_codes', 'get_pairwise_sequence_identity', 'get_sequence_identity', 'get_symbols', 'matrix', 'multiple', 'pairwise', 'score']\n",
      "    A   C   G   T\n",
      "A   5  -4  -4  -4\n",
      "C  -4   5  -4  -4\n",
      "G  -4  -4   5  -4\n",
      "T  -4  -4  -4   5\n",
      "    A   C   D   E   F   G   H   I   K   L   M   N   P   Q   R   S   T   V   W   Y   B   Z   X   *\n",
      "A   4   0  -2  -1  -2   0  -2  -1  -1  -1  -1  -2  -1  -1  -1   1   0   0  -3  -2  -2  -1   0  -4\n",
      "C   0   9  -3  -4  -2  -3  -3  -1  -3  -1  -1  -3  -3  -3  -3  -1  -1  -1  -2  -2  -3  -3  -2  -4\n",
      "D  -2  -3   6   2  -3  -1  -1  -3  -1  -4  -3   1  -1   0  -2   0  -1  -3  -4  -3   4   1  -1  -4\n",
      "E  -1  -4   2   5  -3  -2   0  -3   1  -3  -2   0  -1   2   0   0  -1  -2  -3  -2   1   4  -1  -4\n",
      "F  -2  -2  -3  -3   6  -3  -1   0  -3   0   0  -3  -4  -3  -3  -2  -2  -1   1   3  -3  -3  -1  -4\n",
      "G   0  -3  -1  -2  -3   6  -2  -4  -2  -4  -3   0  -2  -2  -2   0  -2  -3  -2  -3  -1  -2  -1  -4\n",
      "H  -2  -3  -1   0  -1  -2   8  -3  -1  -3  -2   1  -2   0   0  -1  -2  -3  -2   2   0   0  -1  -4\n",
      "I  -1  -1  -3  -3   0  -4  -3   4  -3   2   1  -3  -3  -3  -3  -2  -1   3  -3  -1  -3  -3  -1  -4\n",
      "K  -1  -3  -1   1  -3  -2  -1  -3   5  -2  -1   0  -1   1   2   0  -1  -2  -3  -2   0   1  -1  -4\n",
      "L  -1  -1  -4  -3   0  -4  -3   2  -2   4   2  -3  -3  -2  -2  -2  -1   1  -2  -1  -4  -3  -1  -4\n",
      "M  -1  -1  -3  -2   0  -3  -2   1  -1   2   5  -2  -2   0  -1  -1  -1   1  -1  -1  -3  -1  -1  -4\n",
      "N  -2  -3   1   0  -3   0   1  -3   0  -3  -2   6  -2   0   0   1   0  -3  -4  -2   3   0  -1  -4\n",
      "P  -1  -3  -1  -1  -4  -2  -2  -3  -1  -3  -2  -2   7  -1  -2  -1  -1  -2  -4  -3  -2  -1  -2  -4\n",
      "Q  -1  -3   0   2  -3  -2   0  -3   1  -2   0   0  -1   5   1   0  -1  -2  -2  -1   0   3  -1  -4\n",
      "R  -1  -3  -2   0  -3  -2   0  -3   2  -2  -1   0  -2   1   5  -1  -1  -3  -3  -2  -1   0  -1  -4\n",
      "S   1  -1   0   0  -2   0  -1  -2   0  -2  -1   1  -1   0  -1   4   1  -2  -3  -2   0   0   0  -4\n",
      "T   0  -1  -1  -1  -2  -2  -2  -1  -1  -1  -1   0  -1  -1  -1   1   5   0  -2  -2  -1  -1   0  -4\n",
      "V   0  -1  -3  -2  -1  -3  -3   3  -2   1   1  -3  -2  -2  -3  -2   0   4  -3  -1  -3  -2  -1  -4\n",
      "W  -3  -2  -4  -3   1  -2  -2  -3  -3  -2  -1  -4  -4  -2  -3  -3  -2  -3  11   2  -4  -3  -2  -4\n",
      "Y  -2  -2  -3  -2   3  -3   2  -1  -2  -1  -1  -2  -3  -1  -2  -2  -2  -1   2   7  -3  -2  -1  -4\n",
      "B  -2  -3   4   1  -3  -1   0  -3   0  -4  -3   3  -2   0  -1   0  -1  -3  -4  -3   4   1  -1  -4\n",
      "Z  -1  -3   1   4  -3  -2   0  -3   1  -3  -1   0  -1   3   0   0  -1  -2  -3  -2   1   4  -1  -4\n",
      "X   0  -2  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -2  -1  -1   0   0  -1  -2  -1  -1  -1  -1  -4\n",
      "*  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4  -4   1\n",
      "Load substution matrix from internal database\n",
      "    A   C   D   E   F   G   H   I   K   L   M   N   P   Q   R   S   T   V   W   Y   B   Z   X   *\n",
      "A   5  -1  -2  -1  -3   0  -2  -1  -1  -2  -1  -1  -1  -1  -2   1   0   0  -3  -2  -2  -1  -1  -5\n",
      "C  -1  13  -4  -3  -2  -3  -3  -2  -3  -2  -2  -2  -4  -3  -4  -1  -1  -1  -5  -3  -3  -3  -2  -5\n",
      "D  -2  -4   8   2  -5  -1  -1  -4  -1  -4  -4   2  -1   0  -2   0  -1  -4  -5  -3   5   1  -1  -5\n",
      "E  -1  -3   2   6  -3  -3   0  -4   1  -3  -2   0  -1   2   0  -1  -1  -3  -3  -2   1   5  -1  -5\n",
      "F  -3  -2  -5  -3   8  -4  -1   0  -4   1   0  -4  -4  -4  -3  -3  -2  -1   1   4  -4  -4  -2  -5\n",
      "G   0  -3  -1  -3  -4   8  -2  -4  -2  -4  -3   0  -2  -2  -3   0  -2  -4  -3  -3  -1  -2  -2  -5\n",
      "H  -2  -3  -1   0  -1  -2  10  -4   0  -3  -1   1  -2   1   0  -1  -2  -4  -3   2   0   0  -1  -5\n",
      "I  -1  -2  -4  -4   0  -4  -4   5  -3   2   2  -3  -3  -3  -4  -3  -1   4  -3  -1  -4  -3  -1  -5\n",
      "K  -1  -3  -1   1  -4  -2   0  -3   6  -3  -2   0  -1   2   3   0  -1  -3  -3  -2   0   1  -1  -5\n",
      "L  -2  -2  -4  -3   1  -4  -3   2  -3   5   3  -4  -4  -2  -3  -3  -1   1  -2  -1  -4  -3  -1  -5\n",
      "M  -1  -2  -4  -2   0  -3  -1   2  -2   3   7  -2  -3   0  -2  -2  -1   1  -1   0  -3  -1  -1  -5\n",
      "N  -1  -2   2   0  -4   0   1  -3   0  -4  -2   7  -2   0  -1   1   0  -3  -4  -2   4   0  -1  -5\n",
      "P  -1  -4  -1  -1  -4  -2  -2  -3  -1  -4  -3  -2  10  -1  -3  -1  -1  -3  -4  -3  -2  -1  -2  -5\n",
      "Q  -1  -3   0   2  -4  -2   1  -3   2  -2   0   0  -1   7   1   0  -1  -3  -1  -1   0   4  -1  -5\n",
      "R  -2  -4  -2   0  -3  -3   0  -4   3  -3  -2  -1  -3   1   7  -1  -1  -3  -3  -1  -1   0  -1  -5\n",
      "S   1  -1   0  -1  -3   0  -1  -3   0  -3  -2   1  -1   0  -1   5   2  -2  -4  -2   0   0  -1  -5\n",
      "T   0  -1  -1  -1  -2  -2  -2  -1  -1  -1  -1   0  -1  -1  -1   2   5   0  -3  -2   0  -1   0  -5\n",
      "V   0  -1  -4  -3  -1  -4  -4   4  -3   1   1  -3  -3  -3  -3  -2   0   5  -3  -1  -4  -3  -1  -5\n",
      "W  -3  -5  -5  -3   1  -3  -3  -3  -3  -2  -1  -4  -4  -1  -3  -4  -3  -3  15   2  -5  -2  -3  -5\n",
      "Y  -2  -3  -3  -2   4  -3   2  -1  -2  -1   0  -2  -3  -1  -1  -2  -2  -1   2   8  -3  -2  -1  -5\n",
      "B  -2  -3   5   1  -4  -1   0  -4   0  -4  -3   4  -2   0  -1   0   0  -4  -5  -3   5   2  -1  -5\n",
      "Z  -1  -3   1   5  -4  -2   0  -3   1  -3  -1   0  -1   4   0   0  -1  -3  -2  -2   2   5  -1  -5\n",
      "X  -1  -2  -1  -1  -2  -2  -1  -1  -1  -1  -1  -1  -2  -1  -1  -1   0  -1  -3  -1  -1  -1  -1  -5\n",
      "*  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5  -5   1\n",
      "Local Protein Sequence alignment\n",
      "TITE\n",
      "TITE\n",
      "Global Protein Sequence alignment\n",
      "MITTITE\n",
      "-IT-ITE\n",
      "MITTITE\n",
      "-I-TITE\n",
      "Local Score: 19\n",
      "Local Realc Score: 19\n",
      "Local Sequence Identity: 1.0\n",
      "Global Score: 3\n",
      "Global Realc Score: 3\n",
      "Global Sequence Identity: 0.8333333333333334\n",
      "Find local alignments in nucleotides\n",
      "ATCTCGTA\n",
      "ATCTCGTA\n",
      "Local Score: 40\n",
      "Local Realc Score: 47\n",
      "Local Sequence Identity: 1.0\n",
      "Find global alignments in nucleotides\n",
      "ATGATCTCGTAA\n",
      "AT---CTCGTAG\n",
      "ATGATCTCGTAA\n",
      "A---TCTCGTAG\n",
      "ATGATCTCGTAA\n",
      "---ATCTCGTAG\n",
      "Global Score: 6\n",
      "Global Realc Score: 15\n",
      "Global Sequence Identity: 0.6666666666666666\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAIgAAAAyCAYAAACH65NBAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAABl0lEQVR4nO3aPUoDQRyG8f9u3ECIkCCCQQvdwhsIYucHpLIWPYB4A09iY+kdrJTYWnkEUQshlWIb2KxFYIsQ32gYd2bZ51flA3Ym5GEmYSfK89yAn8S+J4CwEQgkAoFEIJAIBBKBQFpSb/YH6dDM1kqay7Sx+QnY17hex747eolmvT5vMr7iMPP3JflcVYNb0YObEMJCIJDkb5C/Gn1l9nQ5nDz+yCyKI0u6kwZ3r9YtTmZuc849HL/Z4e1mKWOZ+f3c9/1XW06T4nlvv23pWdfZ9Z0G0uw0bO96w8zMnm8+rdGKbeuk43KIIPn83I1mVIz9H9hiIDldQVC+bJTb48V78Tw97VrvoO3s+gRScWwx8IpAILHFVNz0b5DVnZZtn684u36kjhz2BynnEWti0XsxqDkCgUQgkAgEEoFAmhfIuJRZIFjzAmGFqTkCgEQgkDhR5gAnyn6JE2WcKEPNcDe34jhRBoktBl4RCCS2mIrjRBlKwYkyLIRAIBEIJAKBRCCQ5L8YgBUEEoFAIhBIBAKJQCARCKRvxX54UHxAB50AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 144x57.6 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAIgAAAAyCAYAAACH65NBAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAC9UlEQVR4nO3asWsaYRgG8Oe0CqKgNJHGJtA6dFQcAkeHQlshkFDcWhQyBUoHBxfp7B9Rh44lg9JZl0hb6VJIKQRxLbFDg8u1ZCgJqdEOwUCL+S533yte9flBICbw3vvKw/d53meMRiMQXcU36wbI2xgQUmJASIkBISUGhJQYEFK6YfP/PoBbuhd5+nF9ePzb0g3jEAKBfvvg8zAWXNaqIzQPIDOTyPuylz00Jv3drrB2OABA6M0UWe10wwGIzQPIzDTVXYBbDCkxIKRk9xlkIsMwsL29jd3dXQDAYDBAIpGAaZpoNBqO671/8g2PG3fctIKz43N8edm/+P3HOQyfgUDsIvfmq9vwBSZurX+xLAvZbBYA0O/34ff7EY/HAQD7+/sIBoOOepr1PGOtjR4iycDl65WHYSQLMUf9uApIOBxGt9vFyckJQqEQWq0WVldX3ZTSFoz6cf/1xbW/vvkJf8iHu8+ijmosLS3h4OAAAFCpVBCJRFAul8V7vQ6Jecb8QeOylluut5jNzU00m00AQK1WQ6FQ0GqEvMl1QPL5POr1Ok5PT9HpdGCapmRfJOD8bIRPL75f/vQ//HJcw9UWAwDpdBq9Xg+1Wg1bW1tuy9AUSWwxrgMCALlcDuVyGe12G5ZlaTVC3qQVkJ2dHUSjUaRSKbTbbaGWyEu0ArK2toZSqSTVCwkbfwYZW14P4d7zm45qGDZHDkXOI268S0qUEbGXPdSu4aV5pLh9FkMLjgEhJQaElBgQUmJASMkuIEOJixi4/hPIaRuO9Efy0jzTZhcQkRVmJHO3LMJn6I/kpXmmjVsMKTEgpKT1VbsUr5zAkuKVeWZ2osxLJE9gecHcnCijxfDfryB0tX+f5ibzMaw8CjuqwYDMsZlvMdVqFZlMBplMBkdHR1qNeMG8zSNBawUpFosoFotSvczcvM0jgVvMHOOJMhd4omwynigjVxgQUmJASIkBISUGhJTs7mJowXEFISUGhJQYEFJiQEiJASElBoSU/gB1XPT7tjxPpgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 144x57.6 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAIgAAAAyCAYAAACH65NBAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAADTUlEQVR4nO3azUsUYRwH8O8zzugug2uu4aq4vpAWBVEeIogMJPLWoUt16xIePHiKztafEGG3oEMvdAy6BEpkL0REGoVpRRu9sCqN7urqzrq700E0pPXZnZndcdXv57TM8szvYZ8vzzzD/oRlWSDajLLVE6DyxoCQFANCUgwISTEgJMWAkJQq+7J3uD0KIOTi/lk4D+F2HLuVtV3N+8npbyLX9Xw3dBOOQu6/08ZuZe2SPA34iCEpBoSkpGeQzcw8T2B8cBYnbjdBb6kseFwqlsHbq9HVz0YGQhHQ9qxm9PjNJihazsfgBqaRxuSQgfhkCoom4GtQcaA/CL1ZK2ltp3UBwJzLYGrIQGzChFqtQFEF2i4EUH9Szzu2GL+Z0/UCACH7L6Z3uD3nl++vz8A0Mgh2+bDvUq2tgmu+3plDhV9B2/magsdYloU3A1E09uoInw0AABa+mEgvW6g97CtZbTd1c41dnk5j9uUSWs4FCp6zk3mvKWS9nB5S/5NezmL+o4lDV/Yi+jRhd7grc2NJCBXrPzQAVHdU2QqH13WNd0kIbeNYf0i1HQ6n3K6X7YDMvlhC3TE/9GYNWnUF4p9N20WdWoysINBZ5Vm9YtRNfF9BoMP7Oa9xu162AxIdSaChZ/XZ2dCjIzri7S6y3U3c+INXfb/wuv+3J/XcrpetQ2oqloExlsRiJAUIwMoCAkBnXy2EyH9Ycktv1TD9zPtAuqmrt2qYHv039uBAHVKxjCcBKcZ62dpBZkaX0HhGR/e9MLrvhnHqfhi+RhXzH7x5zAS7fMiuWPj5eGH9WuyTCWM8WbZ1g10+ZFMWfjyKr1/LJL1p0irGetnaQaIji2i7uPEEHerWER1OlPygCABCCBwZrMfUkIHIgxiUSgF/SMX+/mDZ1hVC4Oi1ekzeMhB5GEdljYIKn4LOy6WdM1Cc9XL0mks7T9Fec2l3YUBIigEhKQaEpBgQksoXkKwns6CylS8g3GF2OQaApBgQkmJHWYG12VGWAzvK3NdlR5mH2FFmHzvKyrwuO8rYUWYLO8pKiB1l9rCjDOwok2FHGTvKpNhRRkXDjjJyhAEhKQaEpBgQkmJASEr6FkPEHYSkGBCSYkBIigEhKQaEpBgQkvoLKROY3vvp72QAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 144x57.6 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAIgAAAAyCAYAAACH65NBAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAGRUlEQVR4nO2cXUxTZxjH/6dflB1FKfI5GTYq0zFHdeAyNpZoM+LNEtGMkWyzF1uMkWhCQrjZsnW7s8tullWyC5dsySgxGgnJbkzoVMD5EV0xMKZzsw2KhdJikUJbaM8uSGtLy9v3nHbW2Pd3Bec853nOeZ4/55w3/RdOEAQwGKshy/YJMJ5tmEAYRJhAGESYQBhEmEAYRJhAGEQUpJ1N/VongFKKPGFQiu104/XwetUGmljanFRx71+qC3sX3ZmsKyY2K3Eieg0AXLKNqQ6mEQdNnigiTjijcZTiEFNXTGxW4kT0Or1CjNyFCYRBhPgOspKpQR+GjS40/FgB/iVVwv6gN4Qbnc7lnz0hcDIOyvXLGnzj+wrIlE8ec5OTk2hvb8eVK1dQWFgIlUqFzs5ONDc3J6197tw5HDhwAGNjY9i2bVvCfrfbDb1eDwBwOp2Qy+UoLi4GAFy7di0uNuBZwu2THszeDkKm5KAuU+DloxrwG5WSroU2JwAEZkK4c9ID71gAirUyyBQcNn1QgJK3ecm1U82FtofJ4EifxTT1a+N23vp6CgFPCJqdamw2FBIT//PTDOT5MmxqWRe3/bz+HgRBQENDAwwGA44cOQIAcDgc6Ovrw7Fjx5Lma2lpwcOHD6HX62E0Gom1jUYj1qxZg46OjthrAQAIgoDrx50ob+JR+V4BAODx3QCWFgQU7lCLupYItDmTxS1MLsF1eR4vNRdIqg2sPpfz+ntxcSl6KOklNcrSQhiPRgN4pWMDnBd8tIclxWq1QqVSRcUBAFVVVauKY25uDkNDQzh16hR6enrSqj1j84NTIDogAFi7JW9VcWQyp+cPPzhlfFx+qWJVcdBAOxepPaQWiGtoHkX1+eA3KqFcK8fs3wHqIisZHR3Frl27qON7e3uxb98+VFdXQ6PR4ObNm5Jrz9kXUbA1T/Lx6eT0ORZRsCWztWnnIrWH1AJxWn0o27P8nCzbw8NpTe8uEktbWxtqa2tRX1+fdL/FYkFraysAoLW1FRaLJWO1s8nYd278fvgBrh6dkJyDdi5Se0j1khr0huCx+TFnDwIcIISXH1hbDxeC45I+uojU1NTg7Nmz0d/NZjOmp6dRV1eXEOt2u2G1WjEyMgKO4xAKhcBxHEwmk6TafJUSk5cyJ24xOfkqJSYHnsRtP16EoDckWSC0c0mnh1R3kKmBeZS/y6OxuxKNv1TiHUsl1OUKPBqR9pjZu3cv/H4/urq6otvm5+eTxp45cwaHDh2Cw+GA3W7H+Pg4tFotBgcHJdXW7FQjvCjg/q+Po9u8fwXgGfZLyicmp2anGuGggPG+2ei2kF+6YYt2Lun0kEogTuscSt56IW5baSMPZ7+0v0SO49Db24uLFy9Cq9Vi9+7dMBgMOHHiREKsxWJJWPoePHgQ3d3dkmvXGkvgubGAwY/v4/InD/Dvz4+QVySXlE9MTo7joPuqBDO3/Bj46D6utk1g1DSNrZ9qJNWlnUs6PRS1zM0EK5deT4vIMjeXENnr9Ja5jNyECYRBhAmEQYQJhEGECYRBJJVAwpkuGBYynpIKLvlL+nNNJnqdSiAZv8PIuOzctATk3jcIM9Fr9ohhEGECYRDJmqMsm+Sio0ysgy+CKIE4rT6sfzUPzt982GxIPBHVOjne/OFFAHROqGwgCAKGv3ShvInHa5+XAFh2fwVnQnHDFHMttDkFQcDwF1Mob+Kx47NlO2TEURaL2D6mmosgCNi/fz8MBkP085eIgy8VWXGUZZNcdJSJdfDFkhVHWTbJRUeZWAdfLM+EoyxXeZqOslhSOfhioRJIxLn057fTGPhwHPbTXkxe8CHd/05kNpuh0+mg0+kwMUFuEm1sqji+Spnxux9tTr5Kidm7T+K2Hy/C69+UIegNSapLO5eampo4D6rZbEZ/fz9cLlfKGllxlEVoa2uDzWaDzWZDRUVFRmJTxeWio0yMg28lVKsYp3UOm1rj36IjzqV0Xu6yQcT9deekB/YeL2QqDvmlClQflebqEpMz4ii73eWB/fQsVOtkkKtlaTnKaOYScfC1t7fDZDKhuLgYPM8ndfAlXBtzlD2/MEcZ43+HCYRBhAmEQYQJhEGECYRBhLiKYTDYHYRBhAmEQYQJhEGECYRBhAmEQYQJhEHkP82j7G9LEoSCAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 144x57.6 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Similar to Biopython\n",
    "#Local and global alignment\n",
    "#In Biotite need to get a substution matrix then can use alignment\n",
    "\n",
    "import biotite.sequence.align as align\n",
    "import biotite.sequence as seq\n",
    "import biotite.sequence.graphics as graphics\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "print(dir(align))\n",
    "\n",
    "#Create substution matrix, standard matrix, similar to Biopython dotplot, matches give 5 points\n",
    "#mismatches give -4 and gaps give _\n",
    "n_matrix = align.SubstitutionMatrix.std_nucleotide_matrix()\n",
    "print(n_matrix)\n",
    "\n",
    "#Protein matrix\n",
    "p_matrix = align.SubstitutionMatrix.std_protein_matrix()\n",
    "print(p_matrix)\n",
    "\n",
    "#Load substution matrix from internal database\n",
    "print(\"Load substution matrix from internal database\")\n",
    "alpha = seq.ProteinSequence.alphabet\n",
    "matrix_bl50 = align.SubstitutionMatrix(alpha,alpha,\"BLOSUM50\")\n",
    "print(matrix_bl50)\n",
    "\n",
    "#Protein Sequence alignment, csn choose local or global\n",
    "pseq1 = seq.ProteinSequence(\"MITTITE\")\n",
    "pseq2 = seq.ProteinSequence(\"ITITE\")\n",
    "\n",
    "#Local alignment\n",
    "l_alignments = align.align_optimal(pseq1,pseq2,p_matrix,local=True)\n",
    "print(\"Local Protein Sequence alignment\")\n",
    "for a in l_alignments:\n",
    "    print(a)\n",
    "\n",
    "#Global alignment\n",
    "g_alignments = align.align_optimal(pseq1,pseq2,p_matrix,local=False)\n",
    "print(\"Global Protein Sequence alignment\")\n",
    "for a in g_alignments:\n",
    "    print(a)\n",
    "\n",
    "#More fancy representation\n",
    "def fancy_alignment(alignments,matrix_type):\n",
    "    fig,ax = plt.subplots(figsize = (2.0,0.8))\n",
    "    graphics.plot_alignment_similarity_based(\n",
    "        ax,alignments[0], matrix=matrix_type, symbols_per_line=len(alignments[0]))\n",
    "    fig.tight_layout()\n",
    "\n",
    "#Local and global alignment\n",
    "fancy_alignment(l_alignments,p_matrix)\n",
    "fancy_alignment(g_alignments,p_matrix)\n",
    "\n",
    "#Check for local and global alignment scores\n",
    "def check_scores(alignments, align_type):\n",
    "    print(align_type,\"Score:\",alignments.score)\n",
    "    print(align_type,\"Realc Score:\",align.score(alignments,matrix=p_matrix))\n",
    "    print(align_type,\"Sequence Identity:\",align.get_sequence_identity(alignments))\n",
    "\n",
    "check_scores(l_alignments[0], \"Local\")\n",
    "check_scores(g_alignments[0], \"Global\")\n",
    "\n",
    "#Find local and global alignments in nucleotides\n",
    "dna_seq1 = seq.NucleotideSequence(\"ATGATCTCGTAA\")\n",
    "dna_seq2 = seq.NucleotideSequence(\"ATCTCGTAG\")\n",
    "dna__local_alignments = align.align_optimal(dna_seq1,dna_seq2,n_matrix,local=True)\n",
    "dna_global_alignments = align.align_optimal(dna_seq1,dna_seq2,n_matrix,local=False)\n",
    "\n",
    "print(\"Find local alignments in nucleotides\")\n",
    "for a in dna__local_alignments:\n",
    "    print(a)\n",
    "fancy_alignment(dna__local_alignments,n_matrix)\n",
    "check_scores(dna__local_alignments[0], \"Local\")\n",
    "\n",
    "print(\"Find global alignments in nucleotides\")\n",
    "for a in dna_global_alignments:\n",
    "    print(a)\n",
    "fancy_alignment(dna_global_alignments,n_matrix)\n",
    "check_scores(dna_global_alignments[0], \"Global\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
